File: ios/ornacompanion/Model/Specialization.swift
//
//  Item.swift
//  ornacompanion
//
//  Created by Bryan de Ridder on 07/04/2021.
//

import Foundation

struct Specialization: Hashable, Codable, Identifiable {
    var id: Int = 0
    var name: String = ""
    var femaleName: String = ""
    var cost: String = ""
    var price: String = ""
    var description: String = ""
    var tier: Int = 1
    var images: [String] = []
    var skills: Skills = Skills()
    var boosts: [Boost] = []
    
    func getImage() -> String {
        if images.count == 0 {
          return ""
        } else {
          return Constant.ORNA_IMAGE_PREFIX + images[images.count - 1]
        }
    }
    
    func formattedCost() -> String {
        if cost.isEmpty {
            return "\(NumberUtil.formatNumber(price)) orns"
        } else {
            return cost
        }
    }
    
    struct Skills: Hashable, Codable {
        var passives: [Passive] = []
        var learns: [Learn] = []
        var slots: Int = 0
    }
    
    struct Learn: Hashable, Codable, Identifiable {
        var id: Int = 0
        var level: Int = 0
        var name: String = ""
        var type: String = ""
    }
    
    struct Passive: Hashable, Codable, Identifiable {
        var id: Int = 0
        var name: String = ""
    }
    
    struct Boost: Hashable, Codable {
        var name: String
        var value: Int
        
        func formattedName() -> String {
            switch name.lowercased() {
            case "hit points" : return "HP"
            case "attack" : return "Att"
            case "defense" : return "Def"
            case "resistance" : return "Res"
            case "magic" : return "Mag"
            case "dexterity" : return "Dex"
            case "mana" : return "Mana"
            default: return "?"
            }
        }
        
        func formattedValue() -> String {
            if value >= 0 {
                return "+\(value)%"
            } else {
                return "\(value)%"
            }
        }
    }
    
    static let SAMPLE = Specialization(
        id: 56,
        name: "Inquisitor",
        cost: "50,000 orns",
        description: "A long forgotten Inquisitor of Avalon. Learns Holy and defensive skills.",
        tier: 7,
        images: ["subclasses/apprentice/default_m.png"],
        boosts: [Boost(name: "Attack", value: 5), Boost(name: "Hit points", value: -5)]
    )
}

File: android/app/src/main/java/nl/bryanderidder/ornaguide/specialization/model/Specialization.kt
package nl.bryanderidder.ornaguide.specialization.model

import androidx.room.*
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import nl.bryanderidder.ornaguide.characterclass.model.CharacterClass
import nl.bryanderidder.ornaguide.shared.util.NumberUtil
import nl.bryanderidder.ornaguide.shared.util.ORNA_IMAGE_PREFIX
import timber.log.Timber

@Entity
@JsonClass(generateAdapter = true)
data class Specialization(
    @Json(name = "id") @PrimaryKey val id: Int = 0,
    @Json(name = "name") val name: String = "",
    @Json(name = "female_name") val femaleName: String = "",
    @Json(name = "cost") val cost: String = "",
    @Json(name = "price") val price: String = "",
    @Json(name = "description") val description: String = "",
    @Json(name = "tier") val tier: Int = 0,
    @Json(name = "images") val images: List<String> = listOf(),
    @Json(name = "skills") @Embedded val skills: Skills = Skills(),
    @Json(name = "boosts") val boosts: List<Boost> = listOf(),
    @Json(name = "preferred_weapons") val preferredWeapons: List<String> = listOf(),
) {
    @Ignore val imageUrls: List<String> = images.map { ORNA_IMAGE_PREFIX + it }.toList()
    @Ignore val previewImageUrl: String = imageUrls.last()
    @Ignore var isFiltered: Boolean = true

    @JsonClass(generateAdapter = true)
    data class Skills(
        @Json(name = "passives") val passives: List<CharacterClass.Passive> = listOf(),
        @Json(name = "learns") val learns: List<CharacterClass.Learn> = listOf(),
        @Json(name = "slots") val slots: Int = 0
    )

    @JsonClass(generateAdapter = true)
    data class Boost(
        @Json(name = "name") val name: String = "",
        @Json(name = "value") val value: Int = 0
    ) {
        @Ignore
        fun formattedName(): String = when (name.toLowerCase()) {
            "hit points" -> "HP"
            "attack" -> "Att"
            "defense" -> "Def"
            "resistance" -> "Res"
            "magic" -> "Mag"
            "dexterity" -> "Dex"
            "mana" -> "Mana"
            else -> "?"
        }

        @Ignore
        fun formattedValue(): String =
            if (value >= 0)
                "+$value%"
            else
                "$value%"
    }

    @Ignore
    fun formattedCost(): String = try {
        if (cost.isEmpty() && price.isNotEmpty())
            "${NumberUtil.formatNumber(price.toDouble())} orns"
        else
            cost
    } catch (e: NumberFormatException) {
        Timber.e(e, "Couldn't format number for specialization price {\"id\":$id, \"price\": $price}")
        cost
    }

    @Ignore
    fun formattedPreferredWeapons(): String = when {
        preferredWeapons.isEmpty() -> ""
        preferredWeapons.size == 1 -> "Preferred weapon: ${preferredWeapons[0]}"
        else -> "Preferred weapons: ${preferredWeapons.joinToString(", ")}"
    }

    companion object {
        const val NAME = "Specialization"
    }
}

@Entity
@Fts4(contentEntity = Specialization::class)
data class SpecializationFTS(
    val name: String,
)